libr/debug/p/native/windows/windows_debug.c

```diff
@@ -0,0 +1,1164 @@
+		desc = "illegal instruction";
+		break;
+	case EXCEPTION_INT_DIVIDE_BY_ZERO:
+		desc = "divide by zero";
+		break;
+	case EXCEPTION_STACK_OVERFLOW:
+		desc = "stack overflow";
+		break;
+	default:
+		desc = "unknown";
+	}
+
+	return desc;
+}
+
+static int debug_exception_event(DEBUG_EVENT *de) {
+	unsigned long code = de->u.Exception.ExceptionRecord.ExceptionCode;
+	switch (code) {
+	/* fatal exceptions */
+	case EXCEPTION_ACCESS_VIOLATION:
+	case EXCEPTION_ARRAY_BOUNDS_EXCEEDED:
+	case EXCEPTION_ILLEGAL_INSTRUCTION:
+	case EXCEPTION_INT_DIVIDE_BY_ZERO:
+	case EXCEPTION_STACK_OVERFLOW:
+		eprintf ("(%d) Fatal exception (%s) in thread %d\n",
+			(int)de->dwProcessId, 
+			get_w32_excep_name(code),
+			(int)de->dwThreadId);
+		break;
+	/* MS_VC_EXCEPTION */
+	case 0x406D1388:
+		eprintf ("(%d) MS_VC_EXCEPTION (%x) in thread %d\n",
+			(int)de->dwProcessId, (int)code, (int)de->dwThreadId);
+		return 1;
+	default:
+		eprintf ("(%d) Unknown exception %x in thread %d\n",
+			(int)de->dwProcessId, (int)code, (int)de->dwThreadId);
+		break;
+	}
+	return 0;
+}
+
+static char *get_file_name_from_handle(HANDLE handle_file) {
+	HANDLE handle_file_map = NULL;
+	LPTSTR filename = NULL;
+	DWORD file_size_high = 0;
+	LPVOID map = NULL;
+	DWORD file_size_low = GetFileSize (handle_file, &file_size_high);
+
+	if (file_size_low == 0 && file_size_high == 0) {
+		return NULL;
+	}
+	handle_file_map = CreateFileMapping (handle_file, NULL, PAGE_READONLY, 0, 1, NULL);
+	if (!handle_file_map) {
+		goto err_get_file_name_from_handle;
+	}
+	filename = malloc ((MAX_PATH + 1) * sizeof (TCHAR));
+	if (!filename) {
+		goto err_get_file_name_from_handle;
+	}
+	/* Create a file mapping to get the file name. */
+	map = MapViewOfFile (handle_file_map, FILE_MAP_READ, 0, 0, 1);
+	if (!map || !GetMappedFileName (GetCurrentProcess (), map, filename, MAX_PATH)) {
+		goto err_get_file_name_from_handle;
+	}
+	TCHAR temp_buffer[512];
+	/* Translate path with device name to drive letters. */
+	if (!GetLogicalDriveStrings (sizeof (temp_buffer) - 1, temp_buffer)) {
+		goto err_get_file_name_from_handle;
+	}
+	TCHAR name[MAX_PATH];
+	TCHAR drive[3] = TEXT (" :");
+	LPTSTR cur_drive = temp_buffer;
+	while (*cur_drive) {
+		/* Look up each device name */
+		*drive = *cur_drive;
+		if (QueryDosDevice (drive, name, MAX_PATH)) {
+			size_t name_length = _tcslen (name);
+
+			if (name_length < MAX_PATH) {
+				if (_tcsnicmp (filename, name, name_length) == 0
+					&& *(filename + name_length) == TEXT ('\\')) {
+					TCHAR temp_filename[MAX_PATH];
+					_sntprintf (temp_filename, MAX_PATH, TEXT ("%s%s"),
+						drive, filename + name_length);
+					_tcsncpy (filename, temp_filename,
+						_tcslen (temp_filename) + 1);
+					break;
+				}
+			}
+		}
+		cur_drive++;
+	} 
+err_get_file_name_from_handle:
+	if (map) {
+		UnmapViewOfFile (map);
+	}
+	if (handle_file_map) {
+		CloseHandle (handle_file_map);
+	}
+	if (filename) {
+		char *filename_ = r_sys_conv_win_to_utf8 (filename);
```

Comment:
use better varname like utf8filename

Reply:
Once again this code wasn't mine and only had a few touch ups to fix bugs or styling.